srtcore/group.cpp

```diff
@@ -1918,6 +1918,218 @@ struct FLookupSocketWithEvent_LOCKED
+    for (gli_t gi = m_Group.begin(); gi != m_Group.end(); ++gi)
+    {
+        if (gi->laststatus == SRTS_CONNECTING)
+        {
+            HCLOG(ds << "@" << gi->id << "<pending> ");
+            continue; // don't read over a failed or pending socket
+        }
+
+        if (gi->laststatus >= SRTS_BROKEN)
+        {
+            broken.insert(gi->ps);
+        }
+
+        if (broken.count(gi->ps))
+        {
+            HCLOG(ds << "@" << gi->id << "<broken> ");
+            continue;
+        }
+
+        if (gi->laststatus != SRTS_CONNECTED)
+        {
+            HCLOG(ds << "@" << gi->id << "<unstable:" << SockStatusStr(gi->laststatus) << "> ");
+            // Sockets in this state are ignored. We are waiting until it
+            // achieves CONNECTING state, then it's added to write.
+            // Or gets broken and closed in the next step.
+            continue;
+        }
+
+        // Don't skip packets that are ahead because if we have a situation
+        // that all links are either "elephants" (do not report read readiness)
+        // and "kangaroos" (have already delivered an ahead packet) then
+        // omiting kangaroos will result in only elephants to be polled for
+        // reading. Due to the strict timing requirements and ensurance that
+        // TSBPD on every link will result in exactly the same delivery time
+        // for a packet of given sequence, having an elephant and kangaroo in
+        // one cage means that the elephant is simply a broken or half-broken
+        // link (the data are not delivered, but it will get repaired soon,
+        // enough for SRT to maintain the connection, but it will still drop
+        // packets that didn't arrive in time), in both cases it may
+        // potentially block the reading for an indefinite time, while
+        // simultaneously a kangaroo might be a link that got some packets
+        // dropped, but then it's still capable to deliver packets on time.
+
+        // Note that gi->id might be a socket that was previously being polled
+        // on write, when it's attempting to connect, but now it's connected.
+        // This will update the socket with the new event set.
+
+        aliveMembers.push_back(gi->ps);
+        HCLOG(ds << "@" << gi->id << "[READ] ");
+    }
+
+    HLOGC(grlog.Debug, log << "group/recv: " << ds.str() << " --> EPOLL/SWAIT");
+#undef HCLOG
+
+    return aliveMembers;
+}
+
+vector<CUDTSocket*> CUDTGroup::recv_CollectReadReady(set<CUDTSocket*>& broken)
+{
+    // TODO: Function. Get a list of readable (non-broken) sockets.
+    {
+        // You can't lock the whole group for that
+        // action because this will result in a deadlock.
+        // Prepare first the list of sockets to be added as connect-pending
+        // and as read-ready, then unlock the group, and then add them to epoll.
+        vector<CUDTSocket*> aliveMembers = recv_CollectNonBroken(broken);
```

Comment:
Chaining like that makes that effectively the `recv_CollectReadReady` call performs two unrelated activities. I can't see why this function couldn't be simply called from within receiveBackup.

Reply:
I think it would be also nice to see directly in the sender function in which call you're going to only collect data - and it is expecte to be fast and do nothing with burden - and in which you're going to wait for read-ready links - which involves calling the waiting function with a longer time to wait, as well as unlock and lock back a mutex.

Reply:
Two functions now:
- `recv_CollectAliveAndBroken(vector& alive, set& broken)` - classifies broken and alive links, fills corresponding containers.
- `vector<CUDTSocket*> recv_WaitForReadReady(const vector& alive, set& broken)`  - adds alive sockets to epoll and waits for read-ready notification. Returns sockets that signall read-readiness. Expecting RVO as vector holds pointers.